internal/host/plugin/repository_host_set.go

```diff
@@ -152,6 +156,364 @@ func (r *Repository) CreateSet(ctx context.Context, scopeId string, s *HostSet,
+	// we use for the plugin hook calls.
+	plgHc, err := toPluginCatalog(ctx, catalog)
+	if err != nil {
+		return nil, nil, nil, db.NoRowsAffected, errors.Wrap(ctx, err, op)
+	}
+	currentPlgSet, err := toPluginSet(ctx, currentSet)
+	if err != nil {
+		return nil, nil, nil, db.NoRowsAffected, errors.Wrap(ctx, err, op)
+	}
+	newPlgSet, err := toPluginSet(ctx, newSet)
+	if err != nil {
+		return nil, nil, nil, db.NoRowsAffected, errors.Wrap(ctx, err, op)
+	}
+
+	// Get the preferred endpoints to write out.
+	var preferredEndpoints []interface{}
+	if endpointOp == endpointOpUpdate {
+		preferredEndpoints = make([]interface{}, 0, len(newSet.PreferredEndpoints))
+		for i, e := range newSet.PreferredEndpoints {
+			obj, err := host.NewPreferredEndpoint(ctx, newSet.PublicId, uint32(i+1), e)
+			if err != nil {
+				return nil, nil, nil, db.NoRowsAffected, errors.Wrap(ctx, err, op)
+			}
+			preferredEndpoints = append(preferredEndpoints, obj)
+		}
+	}
+
+	oplogWrapper, err := r.kms.GetWrapper(ctx, scopeId, kms.KeyPurposeOplog)
+	if err != nil {
+		return nil, nil, nil, db.NoRowsAffected, errors.Wrap(ctx, err, op, errors.WithMsg("unable to get oplog wrapper"))
+	}
+
+	// If the call to the plugin succeeded, we do not want to call it again if
+	// the transaction failed and is being retried.
+	var pluginCalledSuccessfully bool
+
+	var recordUpdated bool
+	var returnedSet *HostSet
+	var hosts []*Host
+	_, err = r.writer.DoTx(
+		ctx,
+		db.StdRetryCnt,
+		db.ExpBackoff{},
+		func(r db.Reader, w db.Writer) error {
+			msgs := make([]*oplog.Message, 0, len(preferredEndpoints)+2)
+			ticket, err := w.GetTicket(s)
+			if err != nil {
+				return errors.Wrap(ctx, err, op, errors.WithMsg("unable to get ticket"))
+			}
+
+			if len(dbMask) != 0 || len(nullFields) != 0 {
+				returnedSet = newSet.clone()
+				var hsOplogMsg oplog.Message
+				numUpdated, err := w.Update(
+					ctx,
+					returnedSet,
+					dbMask,
+					nullFields,
+					db.NewOplogMsg(&hsOplogMsg),
+					db.WithVersion(&version),
+				)
+				if err != nil {
+					return errors.Wrap(ctx, err, op)
+				}
+
+				if numUpdated != 1 {
+					return errors.New(ctx, errors.MultipleRecords, op, fmt.Sprintf("expected 1 set to be updated, got %d", numUpdated))
+				}
+
+				recordUpdated = true
+				msgs = append(msgs, &hsOplogMsg)
+			}
+
+			switch endpointOp {
+			case endpointOpDelete:
+				// Delete all old endpoint entries.
+				var peDeleteOplogMsg oplog.Message
+				pep := &host.PreferredEndpoint{
+					PreferredEndpoint: &store.PreferredEndpoint{
+						HostSetId: newSet.PublicId,
+						Priority:  1,
+					},
+				}
+				_, err := w.Delete(ctx, pep, db.WithWhere("host_set_id = ?", newSet.PublicId), db.NewOplogMsg(&peDeleteOplogMsg))
+				if err != nil {
+					return errors.Wrap(ctx, err, op)
+				}
+
+				// Only append the oplog message if an operation was actually
+				// performed.
+				if peDeleteOplogMsg.Message != nil {
+					if !recordUpdated {
```

Comment:
instead of updating the version buried deep in these if statements, keep track of whether the record was updated (like you are with `recordUpdated` and if the preferred endpoints were updated (with a new bool) and update the version accordingly at the end.

Reply:
Done!